course/api/views.py

```diff
@@ -64,6 +95,112 @@ def get_queryset(self):
     def get_object(self):
         return self.get_member_object('instance', 'Course')
 
+    def set_teachers(self, course: CourseInstance, data: dict) -> None:
+        teachers = []
+        for i in data['teachers']:
+            userprofile: UserProfile = None
+            try:
+                user = User.objects.get(username=i)
+            except User.DoesNotExist:
+                # If user does not exist, create a new user.
+                # If external authentication (e.g. Shibboleth) is used, other
+                # attributes will be updated when user logs in for the first time.
+                user = User.objects.create_user(i)
+
+            userprofile = UserProfile.objects.get(user=user)
+            teachers.append(userprofile)
+
+        course.set_teachers(teachers)
+
+    def create(self, request: HttpRequest, version) -> Response:
+        if (not request.user.is_superuser):
+            return Response(
+                'Only admins can add new course instance.',
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        data = request.data
+        form = CourseInstanceForm(data)
+
+        if not form.is_valid():
+            return Response({'status': 'error', 'errors': form.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        else:
+            code = form.cleaned_data.get('code')
+            course: Course = Course.objects.filter(code=code).first()
+            if not course:
+                try:
+                    course = Course.objects.create(
+                        name = form.cleaned_data.get('name'),
+                        code = code,
+                        url = form.cleaned_data.get('course_url'),
+                    )
+                except:
+                    return Response('Course creation failed.',
+                                    status=status.HTTP_400_BAD_REQUEST)
+
+            formdata = form.cleaned_data
+            instance_name = formdata.get('instance_name')
+            if (not formdata.get('url')):
+                # if url is not given, compose it from instance name, removing special chars
+                formdata['url'] = ''.join(i for i in instance_name if i.isalnum())
```

Comment:
Based on docs (https://docs.python.org/3/library/stdtypes.html#str.isalnum), `isalnum` allows much more than the ASCII printing characters. In the URL (path), should we allow Scandinavian or Chinese letters? Do they actually work? I guess they should work correctly, though they would usually be hex encoded. Is that a problem? Other opinions?

At least, this removes whitespace and special characters (that's good).

Reply:
Decided to give up on building URL automatically, and require that request always includes it. I don't anyone really requested this feature, but it was more an own idea.